spark3/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java

```diff
@@ -0,0 +1,319 @@
+
+      if (format != Format.catalog) {
+        String[] parts = sourceIdent.split("\\.", 2);
+        Path sourcePath = new Path(parts[1]);
+
+        Configuration conf = spark().sessionState().newHadoopConf();
+        FileSystem fs;
+        Boolean isFile;
+        try {
+          fs = sourcePath.getFileSystem(conf);
+          isFile = fs.getFileStatus(sourcePath).isFile();
+        } catch (IOException e) {
+          throw new RuntimeException("Unable to access add_file path", e);
+        }
+
+        if (isFile) {
+          importFile(table, sourcePath, format, partition);
+        } else {
+          importFileTable(table, sourcePath, format, partition);
+        }
+      } else {
+        importCatalogTable(table, sourceIdent, partition);
+      }
+
+      Snapshot snapshot = table.currentSnapshot();
+      long numAddedFiles = Long.parseLong(snapshot.summary().get(SnapshotSummary.ADDED_FILES_PROP));
+      return numAddedFiles;
+    });
+  }
+
+  private static void applyNameMappingIfMissing(Table table) {
+    if (table.properties().get(TableProperties.DEFAULT_NAME_MAPPING) == null) {
+      // Forces Name based resolution instead of position based resolution
+      NameMapping mapping = MappingUtil.create(table.schema());
+      String mappingJson = NameMappingParser.toJson(mapping);
+      table.updateProperties()
+          .set(TableProperties.DEFAULT_NAME_MAPPING, mappingJson)
+          .commit();
+    }
+  }
+
+  private int importFile(Table table, Path file, Format format,  Map<String, String> partition) {
+    if (partition.isEmpty() && !table.spec().isUnpartitioned()) {
+      throw new IllegalArgumentException("Cannot add a file to a partitioned table without specifying the partition " +
+          "which it should be placed in");
+    }
+    if (!partition.isEmpty() && table.spec().isUnpartitioned()) {
+      throw new IllegalArgumentException("Cannot add a file to an unpartitioned table while specifying the partition " +
+          "which it should be placed in");
+    }
+
+    PartitionSpec spec = table.spec();
+    MetricsConfig metricsConfig = MetricsConfig.fromProperties(table.properties());
+    String partitionURI = file.toString();
+    Configuration conf = spark().sessionState().newHadoopConf();
+
+    List<DataFile> files =
+        SparkTableUtil.listPartition(partition, partitionURI, format.name(), spec, conf, metricsConfig);
+
+    if (files.isEmpty()) {
+      throw new IllegalArgumentException(
+          String.format("No file found for add_file command. Looking for a file at URI %s", partitionURI));
+    }
+
+    // Add Snapshot Summary Info?
+    AppendFiles append = table.newAppend();
+    files.forEach(append::appendFile);
+    append.commit();
+    return 1;
+  }
+
+  private void importFileTable(Table table, Path tableLocation, Format format, Map<String, String> partition) {
+    // List Partitions via Spark InMemory file search interface
+    List<SparkTableUtil.SparkPartition> partitions =
```

Comment:
Shall we split it into multiple lines, import `SparkPartition` directly, and rename `partition`?
What about this?

```
List<SparkPartition> partitions = Spark3Util.getPartitions(spark(), location, format.name());
List<SparkPartition> matchingPartitions = filterPartitions(partitions, partitionFilter);
importPartitions(table, matchingPartitions);
```